packages/gatsby/src/datastore/in-memory/run-fast-filters.ts

```diff
@@ -315,51 +317,51 @@ function collectBucketForElemMatch(
  * Filters and sorts a list of nodes using mongodb-like syntax.
  *
  * @param args raw graphql query filter/sort as an object
- * @property {boolean} args.firstOnly true if you want to return only the first
- *   result found. This will return a collection of size 1. Not a single element
- * @property {{filter?: Object, sort?: Object} | undefined} args.queryArgs
+ * @property {{filter?: Object, sort?: Object, skip?: number, limit?: number} | undefined} args.queryArgs
  * @property {FiltersCache} args.filtersCache A cache of indexes where you can
  *   look up Nodes grouped by a FilterCacheKey, which yields a Map which holds
  *   an arr of Nodes for the value that the filter is trying to query against.
  *   This object lives in query/query-runner.js and is passed down runQuery.
- * @returns Collection of results. Collection will be limited to 1
- *   if `firstOnly` is true
+ * @returns Collection of results. Collection will be sliced by `skip` and `limit`
  */
-export function runFastFiltersAndSort(
-  args: IRunFilterArg
-): Array<IGatsbyNode> | null {
+export function runFastFiltersAndSort(args: IRunFilterArg): IQueryResult {
   const {
-    queryArgs: { filter, sort } = {},
+    queryArgs: { filter, sort, limit, skip = 0 } = {},
     resolvedFields = {},
-    firstOnly = false,
     nodeTypeNames,
     filtersCache,
     stats,
   } = args
 
   const result = convertAndApplyFastFilters(
     filter,
-    firstOnly,
     nodeTypeNames,
     filtersCache,
     resolvedFields,
     stats
   )
 
-  return sortNodes(result, sort, resolvedFields, stats)
+  const sortedResult = sortNodes(result, sort, resolvedFields, stats)
+  const totalCount = sortedResult.length
+
+  const entries =
+    skip || limit
+      ? sortedResult.slice(skip, limit ? skip + (limit ?? 0) : undefined)
+      : sortedResult
+
+  return { entries, totalCount }
```

Comment:
The key change: applying `skip`/`limit` in the datastore itself. And providing `totalCount` separately.